proxy/proxy/client_test.go

```diff
@@ -122,3 +124,59 @@ func TestConcurrentRefresh(t *testing.T) {
 	}
 	b.Unlock()
 }
+
+func TestMaximumConnectionsCount(t *testing.T) {
+	const maxConnections = 10
+	const numConnections = maxConnections + 1
+	var dials uint64 = 0
+
+	b := &fakeCerts{}
+	certSource := blockingCertSource{
+		map[string]*fakeCerts{}}
+	c := &Client{
+		Certs: &certSource,
+		Dialer: func(string, string) (net.Conn, error) {
+			atomic.AddUint64(&dials, 1)
+
+			// Wait a second to ensure the max connections count is reached by concurrent dialers
+			time.Sleep(time.Second)
```

Comment:
Sleeping in a test is just asking for a flaky test in the long run unfortunately. I think the Dial only needs to block until after the first failure to dial, i.e. dials can unblock after the first `c.handleConn` returns. Feel free to pick how you implement, but I suggest something like:

    firstDialExited := make(chan struct{})
    c := &Client{
       ...,
       Dialer: func(string, string) (net.Conn, error) {
          atomic.AddUint64(&dials, 1)
          <-firstDialExited
          return nil, errFakeDial
       },
    }
    ...
    var wg sync.WaitGroup
    var firstDialOnce sync.Once
    for _, instanceName := range instanceNames {
       wg.Add(1)
       go func(instanceName string) {
          defer wg.Done()

          conn := Conn{...}
          c.handleConn(conn)

          firstDialOnce.Do(func() { close(firstDialExited) }
       }(instanceName)
    }

wdyt?